src/fluree/db/conn_events.cljc

```diff
@@ -0,0 +1,63 @@
+(ns fluree.db.conn-events
+  (:require [fluree.db.util.core :as util :refer [try* catch*]]
+            [fluree.db.util.log :as log]
+            #?(:clj  [clojure.core.async :as async]
+               :cljs [cljs.core.async :as async])))
+
+(defmulti process-event (fn [_ event-type _ _] event-type))
+
+;; internal, connection-specific events to capture and process
+
+(defmethod process-event :set-ws-id
+  [conn _ _ ws-id]
+  (swap! (:state conn) assoc :socket-id ws-id))
+
+
+(defmethod process-event :default
+  [conn event-type subject event-data]
+  ;; any event not explicitly captured above is assumed to be a 'global' event.
+  ;; send to any registered callback functions
+  (let [callbacks (get-in @(:state conn) [:listeners subject])]
+    (doseq [[k f] callbacks]
+      (#?(:clj future :cljs do)
+        (try* (f event-type event-data)
+              (catch* e
+                      (log/error e
+                                 (str "Error calling registered callback: " (pr-str k) " for db: " subject
+                                      ". Event: " event-type " Data: " (pr-str event-data) "."))))))))
+
+
+(defn process-events
+  "Processes incoming events from the ledger.
+
+  Requests that carry a :req-id in the header will attempt to lookup a
+  corresponding return channel and send the message along on it.
+
+  Otherwise looks for database events and updates local state accordingly."
+  [conn msg]
+  (try*
+    (let [_          (log/trace "Process events: " msg)
```

Comment:
The `log/*` fns should insert spaces between args for you (like `println`), so you can remove the one at the end of the string.